flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobSubmitHandler.java

```diff
@@ -54,18 +69,89 @@ public JobSubmitHandler(
 
 	@Override
 	protected CompletableFuture<JobSubmitResponseBody> handleRequest(@Nonnull HandlerRequest<JobSubmitRequestBody, EmptyMessageParameters> request, @Nonnull DispatcherGateway gateway) throws RestHandlerException {
-		JobGraph jobGraph;
-		try {
-			ObjectInputStream objectIn = new ObjectInputStream(new ByteArrayInputStream(request.getRequestBody().serializedJobGraph));
-			jobGraph = (JobGraph) objectIn.readObject();
-		} catch (Exception e) {
-			throw new RestHandlerException(
-				"Failed to deserialize JobGraph.",
-				HttpResponseStatus.BAD_REQUEST,
-				e);
+		Collection<Path> uploadedFiles = request.getUploadedFiles();
```

Comment:
Could this also return a `Collection<File>`? I think there would be two benefits: 1. indicating that this file is really a local file and 2. we would not have the problem with two `Path` classes in the same scope.

Reply:
I was wondering when you might bring that up :)

I'm not a fan of exposing `File`.
* whether it _really is a local file_ shouldn't be relevant to the handler
* nio Paths are more flexible than files; for example if the DistributedCache were to return files we wouldn't have to extract zips as we could mount it with a `ZipFileSystem` instead
* always having to convert using `File#toPath` is rather tedious

Reply:
I agree that nio `Paths` are more powerful. I'm just wondering whether we actually need this flexibility here. Usually it is a good idea to make the interface as restrictive as possible and widen it on demand.

Moreover, there is also Flink's `Path` which comes with a bit of different semantics. For example, you have the safety net for closing open file streams which is also used to interrupt I/O operations which are otherwise not interruptible. Mixing the nio paths in there, might give the false impression that this also applies to them.

So I'm just saying that we should make this decision consciously and not only based on which type is more convenient to be used.

Reply:
So how far do you propose to go? Should `HandlerRequest` also receive a `Collection<File>`, or convert the existing `Collection<Path>`? Should `FileUploads#getUploadedFiles` return a `Collection<File>`?

Reply:
What about `ClientUtils`, should they also accept `File` collections?